pkg/bmc/bmc.go

```diff
@@ -347,3 +367,199 @@ func redfishGetSystemSecureBoot(redfishClient *gofish.APIClient, systemIndex int
+	if combineOutput {
+		return string(combinedOutput), "", nil
+	}
+
+	return stdoutBuffer.String(), stderrBuffer.String(), nil
+}
+
+// OpenSerialConsole opens the serial console port. The console is tunneled in an underlying (CLI)
+// ssh session that is opened in the BMC's ssh server. If openConsoleCliCmd is
+// provided, it will be sent to the BMC's cli. Otherwise, a best effort will
+// be made to run the appropriate cli command based on the system manufacturer.
+func (bmc *BMC) OpenSerialConsole(openConsoleCliCmd string) (io.Reader, io.WriteCloser, error) {
+	glog.V(100).Infof("Opening serial console.")
+
+	if bmc.sshSessionForSerialConsole != nil {
+		glog.V(100).Infof("There is already a serial console opened for this BMC. Use OpenSerialConsole() first.")
+
+		return nil, nil, fmt.Errorf("there is already a serial console opened for this BMC")
+	}
+
+	cliCmd := openConsoleCliCmd
+	if cliCmd == "" {
+		// no cli command to get console port was provided, try to guess based on
+		// manufacturer.
+		manufacturer, err := bmc.SystemManufacturer()
+		if err != nil {
+			glog.V(100).Infof("Failed to get redifsh system manufacturer: %v", err)
+
+			return nil, nil, fmt.Errorf("failed to get redfish system manufacturer: %w", err)
+		}
+
+		var found bool
+		if cliCmd, found = cliCmdSerialConsole[manufacturer]; !found {
+			glog.V(100).Infof("CLI command to get serial console not found for manufacturer %v", manufacturer)
+
+			return nil, nil, fmt.Errorf("cli command to get serial console not found for manufacturer %v", manufacturer)
+		}
+	}
+
+	sshSession, err := bmc.CreateCLISSHSession()
+	if err != nil {
+		glog.V(100).Infof("Failed to create underlying ssh session: %v", err)
+
+		return nil, nil, err
+	}
+
+	// Pipes need to be retrieved before session.Start()
+	reader, err := sshSession.StdoutPipe()
+	if err != nil {
+		glog.V(100).Infof("Failed to get stdout pipe from ssh session: %v", err)
+
+		_ = sshSession.Close()
+
+		return nil, nil, fmt.Errorf("failed to get stdout pipe from ssh session: %w", err)
+	}
+
+	writer, err := sshSession.StdinPipe()
+	if err != nil {
+		glog.V(100).Infof("Failed to get stdin pipe from ssh session: %v", err)
+
+		_ = sshSession.Close()
+
+		return nil, nil, fmt.Errorf("failed to get stdin pipe from ssh session: %w", err)
+	}
+
+	err = sshSession.Start(cliCmd)
+	if err != nil {
+		glog.V(100).Infof("Failed to start CLI command %q: %v", cliCmd, err)
+
+		_ = sshSession.Close()
+
+		return nil, nil, fmt.Errorf("failed to start serial console with cli command %q: %w", cliCmd, err)
+	}
+
+	go func() { _ = sshSession.Wait() }()
+
+	bmc.sshSessionForSerialConsole = sshSession
+
+	return reader, writer, nil
+}
+
+// CloseSerialConsole closes the serial console's underlying ssh session.
+func (bmc *BMC) CloseSerialConsole() error {
+	glog.V(100).Infof("Closing serial console.")
+
+	if bmc.sshSessionForSerialConsole == nil {
+		glog.V(100).Infof("No underlying ssh session found. Please use OpenSerialConsole() first.")
```

Comment:
```suggestion
		glog.V(100).Infof("No underlying ssh session found for %s. Please use OpenSerialConsole() first.", bmc.host)
```

Reply:
Done in https://github.com/openshift-kni/eco-goinfra/pull/419/commits/c9a7fff0339cd048ad3ab836a5afd4b0cbb6fdb7